### PR TITLE
Rearchitecting the Sharding Node, its Lifecycle, and Services

### PR DESCRIPTION
**Pull request by rauljordan**
Hi all,

This is a PR referencing the recent issues #126 #122. The idea is to make our sharding implementation have a single entry point that kicks off an independent node responsible for handling services and protocols related to sharding. Specifically, we consider `Notary` and `Proposer` to be protocols that can be registered to a sharding node upon startup, alongside shardp2p, and SMC contract bindings. This design is meant to abstract responsibilities into a more modular fashion and make the node be responsible for the lifecycle of each service, which is something each service should not be concerned with.

## Design Rationale

TODO: rauljordan

## Requirements for Merge

- [x] All tests passing
- [x] Notary/Proposer services as they used to before but with the new architecture
- [x] Allow for sharding node to manage lifecycle of notaries and proposers
- [x] A single place for configuration of this sharding node dependent on cli flags

### CODE REVIEW

File: sharding/proposer/service.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+func (p *Proposer) Start() error {
+	log.Info("Starting proposer service")
+	// TODO: Propose collations.
+	return nil
+}
+
+// Stop the main loop for proposing collations.
+func (p *Proposer) Stop() error {
+	log.Info("Stopping proposer service")
+	// TODO: Propose collations.
```
Comment (prestonvanloon) [inline sharding/proposer/service.go]: Why

File: sharding/notary/service.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+		}
+	}
+
+	return subscribeBlockHeaders(n.node)
+}
+
+// Stop the main loop for notarizing collations.
+func (n *Notary) Stop() error {
+	log.Info("Stopping notary service")
+	// TODO: Propose collations.
```
Comment (prestonvanloon) [inline sharding/notary/service.go]: why

File: cmd/geth/shardingcmd.go [hunk 3]
```diff
@@ -1,46 +1,59 @@
...
+// for sharding.
+func registerShardingServices(n node.Node) error {
+	protocolFlag := n.Context().GlobalString(utils.ProtocolFlag.Name)
+
+	err := n.Register(func(ctx *cli.Context) (sharding.Service, error) {
+		// TODO(terenc3t): handle case when we just want to start an observer node.
+		if protocolFlag == "notary" {
+			return notary.NewNotary(n.Context(), n)
+		}
+		return proposer.NewProposer(n.Context(), n)
```
Comment (prestonvanloon) [inline cmd/geth/shardingcmd.go]: if (protocolFlag == "proposer") ? 

The default behavior would be just an observer right?

File: cmd/utils/flags.go [hunk 4]
```diff
@@ -534,10 +534,14 @@ var (
...
-	//Sharding Settings
+	// Sharding Settings
 	DepositFlag = cli.BoolFlag{
 		Name:  "deposit",
-		Usage: "To become a notary with your sharding client, " + new(big.Int).Div(sharding.NotaryDeposit, new(big.Int).Exp(big.NewInt(10), big.NewInt(18), nil)).String() + " ETH will be deposited into SMC",
+		Usage: "To become a notary in a sharding node, " + new(big.Int).Div(sharding.NotaryDeposit, new(big.Int).Exp(big.NewInt(10), big.NewInt(18), nil)).String() + " ETH will be deposited into SMC",
+	}
+	ProtocolFlag = cli.StringFlag{
+		Name:  "protocol",
+		Usage: "notary | proposer",
```
Comment (prestonvanloon) [inline cmd/utils/flags.go]: Could you add more of a description than this?

File: cmd/utils/flags.go [hunk 5]
```diff
@@ -534,10 +534,14 @@ var (
...
 
-	//Sharding Settings
+	// Sharding Settings
 	DepositFlag = cli.BoolFlag{
 		Name:  "deposit",
-		Usage: "To become a notary with your sharding client, " + new(big.Int).Div(sharding.NotaryDeposit, new(big.Int).Exp(big.NewInt(10), big.NewInt(18), nil)).String() + " ETH will be deposited into SMC",
+		Usage: "To become a notary in a sharding node, " + new(big.Int).Div(sharding.NotaryDeposit, new(big.Int).Exp(big.NewInt(10), big.NewInt(18), nil)).String() + " ETH will be deposited into SMC",
+	}
+	ProtocolFlag = cli.StringFlag{
+		Name:  "protocol",
```
Comment (prestonvanloon) [inline cmd/utils/flags.go]: Is protocol the correct flag name? What about `actor` or `role`?

File: cmd/geth/shardingcmd.go (hunk 3 above)
Comment (rauljordan) [reply cmd/geth/shardingcmd.go]: Yeah, added a TODO for default behavior assigned to @terenc3t if the flag is not passed in

File: cmd/utils/flags.go (hunk 5 above)
Comment (rauljordan) [reply cmd/utils/flags.go]: Actor sounds good to me

**Comment by nisdas** [comment]
LGTM

File: cmd/geth/shardingcmd.go [hunk 6]
```diff
@@ -1,46 +1,63 @@
...
-func notaryClient(ctx *cli.Context) error {
-	c := notary.NewNotary(ctx)
-	return c.Start()
+// sharding is the main cmd line entry point for starting a sharding-enabled node.
+// A sharding node launches a suite of services including notary services,
+// proposer services, and a shardp2p protocol.
+func sharding(ctx *cli.Context) error {
+	// configures a sharding-enabled node using the cli's context.
+	shardingNode := node.NewNode(ctx)
+	registerShardingServices(shardingNode)
```
Comment (terenc3t) [inline cmd/geth/shardingcmd.go]: `registerShardingServices` returns an error. Do we want to check for `if err != nil`?

File: cmd/geth/shardingcmd.go [hunk 7]
```diff
@@ -1,46 +1,63 @@
...
-	p := proposer.NewProposer(ctx)
-	return p.Start()
+// registerShardingServices sets up either a notary or proposer
+// sharding service dependent on the ClientType cli flag. We should be defining
+// the services we want to register here, as this is the geth command entry point
+// for sharding.
+func registerShardingServices(n node.Node) error {
+	actorFlag := n.Context().GlobalString(utils.ActorFlag.Name)
+
+	err := n.Register(func(ctx *cli.Context) (sharding.Service, error) {
```
Comment (terenc3t) [inline cmd/geth/shardingcmd.go]: `ctx` is not used here

File: sharding/notary/service.go [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+// Notary holds functionality required to run a collation notary
+// in a sharded system. Must satisfy the Service interface defined in
+// sharding/service.go.
+type Notary struct {
+	node node.Node
+}
+
+// NewNotary creates a new notary instance.
+func NewNotary(ctx *cli.Context, node node.Node) (*Notary, error) {
```
Comment (terenc3t) [inline sharding/notary/service.go]: ctx is unused here, same for `NewProposer`

File: sharding/node/node.go [hunk 9]
```diff
@@ -103,7 +103,7 @@ func (n *shardingNode) Start() error {
 	// Starts every service attached to the sharding node.
 	for _, serviceFunc := range n.serviceFuncs {
 		// Initializes each service by passing in the node's cli context.
```
Comment (terenc3t) [inline sharding/node/node.go]: // Initializes each service ~~by passing in the node's cli context~~ 
We are not passing in the cli context anymore

File: sharding/node/utils.go [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+	b, err := n.client.CodeAt(context.Background(), sharding.ShardingManagerAddress, nil)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get contract code at %s: %v", sharding.ShardingManagerAddress, err)
+	}
+
+	// Deploy SMC for development only.
+	// TODO: Separate contract deployment from the sharding node. It would only need to be deployed
+	// once on the mainnet, so this code would not need to ship with the node.
+	if len(b) == 0 {
+		log.Info(fmt.Sprintf("No sharding manager contract found at %s. Deploying new contract.", sharding.ShardingManagerAddress.String()))
```
Comment (prestonvanloon) [inline sharding/node/utils.go]: Please do not call `String()`

File: sharding/node/utils.go [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+
+		txOps, err := n.CreateTXOpts(big.NewInt(0))
+		if err != nil {
+			return nil, fmt.Errorf("unable to intiate the transaction: %v", err)
```
Comment (prestonvanloon) [inline sharding/node/utils.go]: Typo

File: sharding/node/utils.go [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+		}
+
+		addr, tx, contract, err := contracts.DeploySMC(txOps, n.client)
+		if err != nil {
+			return nil, fmt.Errorf("unable to deploy sharding manager contract: %v", err)
+		}
+
+		for pending := true; pending; _, pending, err = n.client.TransactionByHash(context.Background(), tx.Hash()) {
```
Comment (prestonvanloon) [inline sharding/node/utils.go]: No need for the pending conditional to make a while loop
```go
		for {
			if err = n.client.TransactionByHash(context.Background(), tx.Hash()); err != nil {
				return nil, fmt.Errorf("unable to get transaction by hash: %v", err)
			}
			time.Sleep(1 * time.Second)
		}
```

File: sharding/node/utils.go [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+			if err != nil {
+				return nil, fmt.Errorf("unable to get transaction by hash: %v", err)
+			}
+			time.Sleep(1 * time.Second)
+		}
+
+		log.Info(fmt.Sprintf("New contract deployed at %s", addr.String()))
```
Comment (prestonvanloon) [inline sharding/node/utils.go]: Do you need to call `String()`? The formatter should call this.

File: sharding/node/node.go [hunk 14]
```diff
@@ -0,0 +1,247 @@
...
+// General node for a sharding-enabled system.
+// Communicates to Geth node via JSON RPC.
+type shardingNode struct {
+	endpoint     string                        // Endpoint to JSON RPC.
+	client       *ethclient.Client             // Ethereum RPC client.
+	keystore     *keystore.KeyStore            // Keystore containing the single signer.
+	ctx          *cli.Context                  // Command line context.
+	smc          *contracts.SMC                // The deployed sharding management contract.
+	rpcClient    *rpc.Client                   // The RPC client connection to the main geth node.
+	lock         sync.RWMutex                  // Mutex lock for concurrency management.
```
[the remark this answers is not in the document]
Comment (rauljordan) [reply sharding/node/node.go]: Im assuming the register function can be called in various scenarios concurrently, and I'd rather take care of potential concurrency problems now. I'll replace with a simple sync.Mutex.

File: sharding/node/node.go [hunk 15]
```diff
@@ -0,0 +1,247 @@
...
+
+// Account to use for sharding transactions.
+func (n *shardingNode) Account() *accounts.Account {
+	accounts := n.keystore.Accounts()
+
+	return &accounts[0]
+}
+
+// Context returns the cli context.
+func (n *shardingNode) Context() *cli.Context {
```
[the remark this answers is not in the document]
Comment (rauljordan) [reply sharding/node/node.go]: This occurs in various place where we only have access to a Node interface, and I believe we are passing in the context to several other functions that will use this cli context to configure certain options.

File: sharding/node/utils.go (hunk 12 above)
Comment (rauljordan) [reply sharding/node/utils.go]: Getting: unreachable code right below it

File: sharding/node/node.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (rauljordan) [reply sharding/node/node.go]: Maybe we can explore this as an issue after merge?

File: sharding/node/utils.go (hunk 12 above)
Comment (prestonvanloon) [reply sharding/node/utils.go]: Ok, I misunderstood the code here.

**Pull request merged**